Read PASEF DDA MS2 precursor information

RogerGinBer opened the pull request:
Related to issue #18, I've implemented a function to retrieve PASEF DDA MS2 precursor variables (specifically: `precursorMz`, `precursorCharge`, `precursorIntensity`, `collisionEnergy`, `isolationWindowLowerMz`, `isolationWindowTargetMz`,
 and `isolationWindowUpperMz`) from the TDF tables using `opentimsr`.

This implementation generates a matrix with _all_ this information for the desired input scans, paralleling for all different `fileNames` in the backend. I prefered to create just a single, common function (`.do_calculate_core_ms2_information`) that extracts all variables at once (they are very related and close to each other in the TDF tables) rather than many similar functions for each individual variable: this will make it easy to cache them all at once if we decide to implement cache (#19).

I created methods for all the accessors, as well as unit tests, etc.  AFAIK, we are passing all tests :+1: 

Since this update is pretty significant, I'd suggest we bump the package version to 0.2.0

RogerGinBer committed:
Create internal fun to retrieve MS2 information

RogerGinBer committed:
Create accessor methods, update SpectraData

Create accessor methods and implement MS2 fun into SpectraData.
Create unit tests for methods and check that all works

RogerGinBer committed:
Bump y version and update author list

RogerGinBer committed:
Convert precursorCharge to integer

Ensure that precursorCharge has the proper format, otherwise it would fail as validObject

jorainer commented on R/MsBackendTimsTof.R:
```diff
@@ -288,6 +288,62 @@ setMethod("msLevel", "MsBackendTimsTof", function(object, ...) {
+#' @importMethodsFrom Spectra precursorMz
+#'
+#' @rdname MsBackendTimsTof
+setMethod("precursorMz", "MsBackendTimsTof", function(object, ...) {
```
would it be faster to *just* get the precursor m/z? Just wondering if getting all columns and then subset to precursor m/z might be less efficient.

RogerGinBer replied:
From the profiling tests I've run using a large PASEF DDA file, it seems that matching the scans to the precursor table is by far the longest step, so retrieving one column or all of them wouldn't make that much of a difference: 

![image](https://user-images.githubusercontent.com/45530914/223674679-090612a7-c723-41af-a7ff-c276dc6c6333.png)

I'm working on reducing that time to something more reasonable by subsetting in two steps (45s) or using `data.table` key indexing, which cuts it down to 4 sec. Profvis files attached below:

[TimsTofProfiling.zip](https://github.com/rformassspectrometry/MsBackendTimsTof/files/10919247/TimsTofProfiling.zip)

jorainer replied:
Great! Thanks for the performance evaluation!

jorainer replied:
So, we basically keep it the way you have it.

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -357,3 +375,53 @@ MsBackendTimsTof <- function() {
+    output <- matrix(NA_real_,
+                     nrow = nrow(indices),
+                     ncol = length(.TIMSTOF_MS2_COLUMNS))
+    tbl <- opentimsr::table2df(tms, c("PASEFFrameMsMsInfo", "Precursors"))
```
do I understand it correctly that all this information is actually present in (SQL?) tables within the original data files?

RogerGinBer replied:
Yep, it's inside the `.tdf` file present in each of the file's `.d` folder 👍

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -357,3 +375,53 @@ MsBackendTimsTof <- function() {
+    }
+    output <- matrix(NA_real_,
+                     nrow = nrow(indices),
+                     ncol = length(.TIMSTOF_MS2_COLUMNS))
```
maybe also add the column names with 

```r
ncol = length(.TIMSTOF_MS2_COLUMNS),
dimnames = list(NULL, .TIMSTOF_MS2_COLUMNS))
```

jorainer commented on R/MsBackendTimsTof.R:
```diff
@@ -288,6 +288,62 @@ setMethod("msLevel", "MsBackendTimsTof", function(object, ...) {
+#' @rdname MsBackendTimsTof
+setMethod("precursorCharge", "MsBackendTimsTof", function(object, ...) {
+    tbl <- .calculate_core_ms2_information(object)
+    tbl[, .TIMSTOF_MS2_COLUMNS == "precursorCharge"]
```
note that `precursorCharge` is expected to return an `integer`, so, you will have to add `as.integer` here.

RogerGinBer replied:
Yep, I missed that one in the last commit

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -357,3 +375,53 @@ MsBackendTimsTof <- function() {
+        row <- tbl[[1]][i, ]
+        prec <- tbl[[2]][row$Precursor,]
+        target_rows <- which(indices[,1] == row$Frame &
+                             MsCoreUtils::between(indices[,2],
```
The `MsCoreUtils::` should not be needed here, since you are importing `between` from `MsCoreUtils`.

jorainer commented on DESCRIPTION:
```diff
@@ -10,6 +10,10 @@ Authors@R:
+      person(given = "Roger", family = "Gine Bertomeu", 
+             email = "roger.gine@iispv.cat",
+             comment = c(ORCID = "0000-0003-0288-9619"),
+             role = "aut"),
```
Please add below in `Imports` `(>= 1.11.2) for `MsCoreUtils` to ensure that the correct version of the package providing this function is used.

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -357,3 +375,53 @@ MsBackendTimsTof <- function() {
+                         prec$Charge,                            ## precursorCharge
+                         prec$Intensity,                         ## precursorIntensity
+                         row$CollisionEnergy,                    ## collisionEnergy
+                         row$IsolationMz - row$IsolationWidth,   ## isolationWindowLowerMz
```
Hm, I'm wondering how the isolation width is defined - just to be sure that it should not be isolationMz - isolationWidth/2 ... can you please check?

jorainer reviewed (changes requested):
Thanks Roger for the PR! I have some comments and change requests. - and sorry for the delay...

jorainer commented:
Please give me a ping once you have pushed changes

RogerGinBer committed:
Refactor MS/MS reading from the TDF tables.

Refactored all of .do_calculate_core_ms2_information because old implementation was very slow. Now we use fast dplyr table joins to extract all relevant infomation from the tables. There's still room for future improvement, probably save the tables in the backendobject when initializing and also filtering them when subsetting the backend. In this way, you could keep the table sizes as small as needed and not have to read the long tables from each file if the backend is only storing a couple spectra.

Added depencencies on dplyr (mandatory for fast table joins) and utils (to automatically fill the `version` slot when creating a new backend object)

Temptative bump to 0.2.0 due to the addition of methods: precursorMz, precursorCharge, etc.

RogerGinBer commented:
Hi there @jorainer ! 

It's been more than a year since I last updated on this PR, but better late than never :sweat_smile: 
So I had a problem with the way `do_calculate_core_ms2_information`  matched the scan (spectra) indices with the `PASEFFrameMsMsInfo` and `Precursors` tables: it was really slow, and didn't scale as well as I wanted for larger datasets.

So I've refactored it in a faster (<3s compared to 30s for a regular LC-IM-MS benchmark file I use) and more elegant way (only returning the desired subset of MS/MS columns the user needs). The only downside is that it uses `dplyr` for doing `left_join` on the tables, so that's one dependency more. On the good side, I'd say `dplyr` is a rather stable dependency to have.

I've run all unit tests again and everything seems to work nicely. 

Related to #18, @chufz , feel free to check it out too and tell me all what you think,

Roger

jorainer commented on R/MsBackendTimsTof.R:
```diff
@@ -137,7 +139,8 @@ setClass("MsBackendTimsTof",
                                fileNames = integer(),
                                readonly = TRUE,
-                               version = "0.1"))
+                               version = as.character(
```
honestly, I'm not totally convinced by this. I know, it would allow us to keep track of the version of the package with which a class was created - but it would not reflect the *version* of the *object*. The version of the object should just change if its definition changes, i.e. if slots were added or removed. I would suggest to keep the old version, but am open to discuss.

RogerGinBer replied:
Now I see what you mean: I thought `version` was meant to be related to the package versioning, not the object _structure_ itself. I see this slot is inherited from the virtual class MsBackend, so it makes sense to keep its definition consistent 

So yes, let's keep the old version then 👍

jorainer reviewed (changes requested):
Great @RogerGinBer ! looks all good - with the exception of the version for the object ;)

chufz commented:
Hi Roger, thanks a lot for this implementation, I can try it out with my data and compare the MS2 spectra to data received in python. Is the current version on your main branch ready for testing?

RogerGinBer commented:
> Hi Roger, thanks a lot for this implementation, I can try it out with my data and compare the MS2 spectra to data received in python. Is the current version on your main branch ready for testing?

Hi @chufz I'd say yes: I designed and ran some tests using some PASEF MS/MS data we acquired a while back and it seems to extract the precursor information correctly. That being said, I haven't had the chance to _cross-validate_ it against another source of data, so please go ahead! 

I'll now revert the changes I did the object `version` slot, but that would not affect your testing

RogerGinBer committed:
Revert changes to 'version' slot

jorainer commented:
@chufz , you should be able to install the version with the changes using `BiocManager::install("RogerGinBer/MsBackendTimsTof")`. Would be great if you could report back whether results are comparable/the same.

@RogerGinBer , seems there are (now?) conflicts with the main branch - can you please try to solve them? otherwise I could also give a hand and provide a PR to your branch? just let me know how you prefer it.
